README.md
[![Build Status](https://travis-ci.org/DestinyItemManager/DIM.svg?branch=master)](https://travis-ci.org/DestinyItemManager/DIM)
[![OpenCollective](https://opencollective.com/dim/backers/badge.svg)](#backers)
[![OpenCollective](https://opencollective.com/dim/sponsors/badge.svg)](#sponsors)

### [Download now from the Chrome Store for free](https://chrome.google.com/webstore/detail/destiny-item-manager/apghicjnekejhfancbkahkhdckhdagna)

## Destiny Item Manager
Destiny Item Manager (DIM) lets [Destiny](http://destinythegame.com/) game players easily move items between their Guardians and the Vault. DIM's goal is to let players equip their guardians quickly. Our Loadouts feature accomplishes this by removing manual steps needed when transferring items.

Loadouts give players the ability to define sets of items that they want on their Guardians. When a loadout is selected, DIM will move all of the items referenced by the Loadout to a Guardian. If the item was equipped by another guardian, the Loadouts feature will replace that item with a similar item, if possible, to allow the Loadout referenced item to be transferred. With a single click of a button, you can have a PVP, PVE, or Raid-ready guardian.

DIM is based on the same services used by the Destiny Companion app to move and equip items. DIM will never ask for your credentials to [Bungie.net](https://www.bungie.net). DIM will not be able to dismantle any of your items.

Visit [/r/DestinyItemManager](http://www.reddit.com/r/DestinyItemManager/) or [@ThisIsDIM](http://twitter.com/ThisIsDIM) for updates and more details.

## Bugs and feature requests
Have a bug or a feature request? Please first search for [existing and closed issues](https://github.com/DestinyItemManager/DIM/issues). If your problem or idea is not addressed yet, please open a new issue.

## Backers
[Support us](https://opencollective.com/dim#backer) with a one-time or monthly donation and help us continue our active development.

<a href="https://opencollective.com/dim/backer/0/website" target="_blank"><img src="https://opencollective.com/dim/backer/0/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/1/website" target="_blank"><img src="https://opencollective.com/dim/backer/1/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/2/website" target="_blank"><img src="https://opencollective.com/dim/backer/2/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/3/website" target="_blank"><img src="https://opencollective.com/dim/backer/3/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/4/website" target="_blank"><img src="https://opencollective.com/dim/backer/4/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/5/website" target="_blank"><img src="https://opencollective.com/dim/backer/5/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/6/website" target="_blank"><img src="https://opencollective.com/dim/backer/6/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/7/website" target="_blank"><img src="https://opencollective.com/dim/backer/7/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/8/website" target="_blank"><img src="https://opencollective.com/dim/backer/8/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/9/website" target="_blank"><img src="https://opencollective.com/dim/backer/9/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/10/website" target="_blank"><img src="https://opencollective.com/dim/backer/10/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/11/website" target="_blank"><img src="https://opencollective.com/dim/backer/11/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/12/website" target="_blank"><img src="https://opencollective.com/dim/backer/12/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/13/website" target="_blank"><img src="https://opencollective.com/dim/backer/13/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/14/website" target="_blank"><img src="https://opencollective.com/dim/backer/14/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/15/website" target="_blank"><img src="https://opencollective.com/dim/backer/15/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/16/website" target="_blank"><img src="https://opencollective.com/dim/backer/16/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/17/website" target="_blank"><img src="https://opencollective.com/dim/backer/17/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/18/website" target="_blank"><img src="https://opencollective.com/dim/backer/18/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/19/website" target="_blank"><img src="https://opencollective.com/dim/backer/19/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/20/website" target="_blank"><img src="https://opencollective.com/dim/backer/20/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/21/website" target="_blank"><img src="https://opencollective.com/dim/backer/21/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/22/website" target="_blank"><img src="https://opencollective.com/dim/backer/22/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/23/website" target="_blank"><img src="https://opencollective.com/dim/backer/23/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/24/website" target="_blank"><img src="https://opencollective.com/dim/backer/24/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/25/website" target="_blank"><img src="https://opencollective.com/dim/backer/25/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/26/website" target="_blank"><img src="https://opencollective.com/dim/backer/26/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/27/website" target="_blank"><img src="https://opencollective.com/dim/backer/27/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/28/website" target="_blank"><img src="https://opencollective.com/dim/backer/28/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/29/website" target="_blank"><img src="https://opencollective.com/dim/backer/29/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/30/website" target="_blank"><img src="https://opencollective.com/dim/backer/30/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/31/website" target="_blank"><img src="https://opencollective.com/dim/backer/31/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/32/website" target="_blank"><img src="https://opencollective.com/dim/backer/32/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/33/website" target="_blank"><img src="https://opencollective.com/dim/backer/33/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/34/website" target="_blank"><img src="https://opencollective.com/dim/backer/34/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/35/website" target="_blank"><img src="https://opencollective.com/dim/backer/35/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/36/website" target="_blank"><img src="https://opencollective.com/dim/backer/36/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/37/website" target="_blank"><img src="https://opencollective.com/dim/backer/37/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/38/website" target="_blank"><img src="https://opencollective.com/dim/backer/38/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/39/website" target="_blank"><img src="https://opencollective.com/dim/backer/39/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/40/website" target="_blank"><img src="https://opencollective.com/dim/backer/40/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/41/website" target="_blank"><img src="https://opencollective.com/dim/backer/41/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/42/website" target="_blank"><img src="https://opencollective.com/dim/backer/42/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/43/website" target="_blank"><img src="https://opencollective.com/dim/backer/43/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/44/website" target="_blank"><img src="https://opencollective.com/dim/backer/44/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/45/website" target="_blank"><img src="https://opencollective.com/dim/backer/45/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/46/website" target="_blank"><img src="https://opencollective.com/dim/backer/46/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/47/website" target="_blank"><img src="https://opencollective.com/dim/backer/47/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/48/website" target="_blank"><img src="https://opencollective.com/dim/backer/48/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/49/website" target="_blank"><img src="https://opencollective.com/dim/backer/49/avatar.svg"></a>
<a href="https://opencollective.com/dim/backer/50/website" target="_blank"><img src="https://opencollective.com/dim/backer/50/avatar.svg"></a>

## Sponsors
Become a sponsor and get your logo here with a link to your site.

<a href="https://opencollective.com/dim/sponsor/0/website" target="_blank"><img src="https://opencollective.com/dim/sponsor/0/avatar.svg"></a>
<a href="https://opencollective.com/dim/sponsor/1/website" target="_blank"><img src="https://opencollective.com/dim/sponsor/1/avatar.svg"></a>
<a href="https://opencollective.com/dim/sponsor/2/website" target="_blank"><img src="https://opencollective.com/dim/sponsor/2/avatar.svg"></a>
<a href="https://opencollective.com/dim/sponsor/3/website" target="_blank"><img src="https://opencollective.com/dim/sponsor/3/avatar.svg"></a>

## Translation
If you speak a language other than English that Destiny supports (Italian, German, French, Spanish, Japanese, or Portuguese), a great way to help with DIM development is to provide translations. See [translation guide](https://github.com/DestinyItemManager/DIM/blob/master/TRANSLATIONS.md) for more info on how to help.

## Developer Quick start
Clone the repo:

* `git clone https://github.com/DestinyItemManager/DIM.git`

Install dependencies:

* Install [NodeJS](https://nodejs.org/en/).
* Run `npm install`
* Windows-based developers will need to install `windows-build-tools` (`npm install --global windows-build-tools`) globally prior to running `npm install`. Refer to issue #1439 for [details](https://github.com/DestinyItemManager/DIM/issues/1439).

Check code Style
* `npm run lint` will tell you if you're following the DIM code style (and automatically fix what it can).

Build in dev mode (extension)
* `npm start`. This will bundle and copy all the assets, as well as watch for changes. You can now run DIM locally by enabling [Chrome Extensions Developer Mode](https://developer.chrome.com/extensions/faq#faq-dev-01) and point to the `dist` folder.

Build in dev mode (website)
* `npm run server`. Visit the URL it prints out to load DIM. It will update automatically as you change files.

Get your own API key:

* Goto [Bungie](https://www.bungie.net/en/Application)
* Open your extension in another tab, copy the url up to `/index.html`.
* Paste your extension url into the `Origin Header` section on bungie.net
* Copy your API-key from bungie.net into DIM developer settings panel when it is loaded.* `npm install`

Check out our [contributor guide](https://github.com/DestinyItemManager/DIM/blob/master/CONTRIBUTING.md) for more tips.

Code released under the [MIT license](http://choosealicense.com/licenses/mit/).


CHANGELOG.md
# Next

* New filters for ornaments - is:ornament, is:ornamentmissing, is:ornamentunlocked
* Fixed a bug where item data would not respect your language settings.
* Weapon reviews now show up immediately, and can be edited.
  - If you have been less than friendly, now would be a very good time to edit yourself and put a better foot forward.
* Logging out now brings you to Bungie's auth page, where you can choose to change account or not.
* Fixed "Clear New Items" not working.
* Adjusted the UI a bunch to make it work better on mobile. Just a start - there's still a long way to go.
* Google Drive syncing is a bit smoother.
* Fixed a case where you couldn't create a new class-specific loadout.
* On Firefox, the new-item shines don't extend past the item anymore.

# 4.3.0

* DIM is now just a website - the extension now just sends you to our website. This gives us one, more cross-platform, place to focus on and enables features we couldn't do with just an extension. Don't forget to import your data from the storage page!
* Scrolling should be smoother overall.
* Vendor weapons now show reviews.
* Add a "sort by name" option for item sorting.
* In Google Chrome (and the next version of Firefox), your local DIM data won't be deleted by the browser in low storage situations if you visit DIM frequently.
* Ratings will no longer disappear from the item details popup the second time it is shown.
* Info popups should do a better job of hiding when you ask them to hide.

# 4.2.4

* Work around a Chrome bug that marked the extension as "corrupted".

# 4.2.3

* Fix log out button.
* Put back the accidentally removed hotkeys for setting tags on items.
* Fixed some visual goofs on Firefox.
* Fix a case where DIM would never finish loading.

# 4.2.2

* Fix DIM being invisible on Firefox
* Fix a case where DIM would never finish loading.
* Put back the accidentally removed hotkeys for setting tags on items.

# 4.2.1

* Actually turn on Google Drive in prod.

# 4.2.0

* Exclude all variants of 'Husk of the Pit' from 'Item Leveling' loadout.
* Add a new storage page (under the floppy disk icon) for managing your DIM data. Import and export to a file, and set up Google Drive storage to sync across machines (website only). You can import your data from the Chrome extension into the website from this page as well.
* The settings page has been cleaned up and reworded.
* Added missing Trials emblems and shaders to the is:trials search.
* DIM should look more like an app if you add it to your home screen on Android.
* DIM will show service alerts from Bungie.

# 4.1.2

* Add a "Log Out" button in settings.

# 4.1.1

* Fixed changelog popup too large to close.

# 4.1.0

* Fixed the logic for deciding which items can be tagged.
* Fix "Make room for postmaster".
* Record books have been moved out of the inventory into their own page. Get a better look at your records, collapse old books, and narrow records down to only those left to complete.
* Fix changing new-item shine, item quality display, and show elemental damage icon preferences. They should apply immediately now, without a reload.x
* Localization updates.
* Fixed objective text in the record book floating above stuff.
* Fixed displaying record objectives that are time-based as time instead of just a number of seconds.
* When pinned to the iOS home screen, DIM now looks more like a regular browser than an app. The upside is you can now actually authorize it when it's pinned!
* Loadouts with a complete set of equipped armor now include a stat bar that will tell you the stat tiers of the equipped loadout pieces.
* Loadouts with non-equipping items now won't *de-equip* those items if they're already equipped. #1567
* The count of items in your loadout is now more accurate.
* DIM is now better at figuring out which platforms you have Destiny accounts on.
* DIM is faster!
* Added Age of Triumph filters is:aot and is:triumph
* Add gunsmith filter is:gunsmith
* Updated filters to remove common items for specific filters (e.g. is:wotm no longer shows exotic items from xur, engrams, and planetary materials)
* Loadout Builder's equip button now operates on the selected character, not your last-played character.
* Loadout Builder no longer has equip and create loadout buttons for loadouts that include vendor items.
* Loadout Builder is faster.
* DIM has a new logo!
* Elemental damage color has been moved to a triangle in the upper-left corner of your weapon.
* See community weapon ratings in DIM, and submit your own! Weapon ratings can be turned on in Settings, and will show up on your individual weapons as well as in the details popup. You can submit your own reviews - each review is specific to the weapon roll you're looking at, so you know whether you've got the god roll.

# v3.17.1

* Fixed a bug with the display of the amount selection controls in the move popup for stackable items.
* Localization updates
* Moved the "VCR" controls for stackable item amount selection to their own row.

# 3.17.0

* Fixed the perk selection in Loadout Builder. #1453
* Integrated Trials-centric weapon reviews (and the ability to rate your own gear (and make comments about your gear)).  Done in conjunction with destinytracker.com.
* Fixed the logic for artifact bonuses to compute the right number. #1477
* Restore some missing images from our build system changes.
* Don't allow engrams to be tagged. #1478
* Add home screen icons (and Safari tab icons, and Windows tile icons) for the website.
* Fixed "is:locked" filters to be consistent for engrams. #1489
* The Beta website is now updated automatically for every PR.
* If you're not logged in to the website, we show the login screen.
* Better error messages for when you have the wrong platform selected, plus the error doesn't cover the platform selector.
* Improved website compatibility with Firefox, Safari, and Edge.
* Many style fixes for Safari.

lib/main.js
var buttons = require('sdk/ui/button/action');
var tabs = require("sdk/tabs");

var button = buttons.ActionButton({
  id: "div-link",
  label: "Destiny Item Viewer",
  icon: {
    "16": "./assets/icon16.png",
    "32": "./assets/icon32.png",
    "64": "./assets/icon64.png"
  },
  onClick: handleClick
});

var data = require("sdk/self").data;
var pageUrl = data.url("window.html");
var pageMod = require("sdk/page-mod");
tabs.on('ready', function(tab) {
	if (tab.url == pageUrl){
		var {Cc, Ci} = require("chrome");
		var cookieValue = "", cookieMgr = Cc["@mozilla.org/cookiemanager;1"].getService(Ci.nsICookieManager);
		
		for (var e = cookieMgr.enumerator; e.hasMoreElements();) {
		  var cookie = e.getNext().QueryInterface(Ci.nsICookie);
		  if (cookie.host.indexOf("bungie.net") > -1){
		  	if (cookie.name == "bungled"){
				cookieValue = cookie.value;
			}	
		  } 
		}
		//console.log("cookieValue: " + cookieValue);
		tab.attach({
		    contentScriptFile: data.url("js/firefox.js"),
			contentScriptWhen: "start",
			contentScriptOptions: {"token" : cookieValue } //referenced as self.options.token
		});	
	}
});

function handleClick(state) {
 	var tab = tabs.open(pageUrl);
}